### ISSUE TITLE
Client: make launched python part of windows job object [Timebox: 3 days]

### ISSUE THREAD

**Issue by FreddieAkeroyd** [ISISComputingGroup/IBEX#7674]
As a user i would not like to have stray python processes left around when the client crashes as this causes matplotlib plotting issues with the websockets. On windows we may be able to resolve this by making sure the python process is part of the same job object as the client, then when the client terminates python will too. This approach is used in procServ to resolve a similar issue with stray subprocesses.  See https://learn.microsoft.com/en-us/windows/win32/procthread/job-objects for a description of job objects

The job object should be created in the client with the `JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE` flag - this means when the final handle on the object is closed all processes terminate. Job object membership is inherited by default, so unless python is spawned in a funny way it should automatically be part of the same job. When the client is killed, this automatically closes the handle on the job object which should then kill python and any other subprocess spawned due to the above flag.

I am not sure how reproducible this is e.g. if you kill the ibex-client.exe from task manager what happens to python.exe started by the scripting window?

To implement this you would likely need to call some windows API functions, this may be possbile via JNA https://github.com/java-native-access/jna or some other java library or we may need to create our own DLL. The basic procedure is similar to that in https://github.com/ISISComputingGroup/EPICS-tools/blob/d4f3827b97e8a9a0bccaa255149dd646dac3477f/procServ-2.6.0/processFactory.cc#L132 but applied to the calling rather than forked process.
 
* call CreateJobObject() to create a job object handle
* add JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE flag to job object using QueryInformationJobObject() and SetInformationJobObject()
* call GetCurrentProcess()   to get handle on current process
* call AssignProcessToJobObject() to associate the process with the job object 
* call CloseHandle() on handle returned from GetCurrentProcess()

However do not call CloseHandle() on handle returned from CreateJobObject() - this will be automatically closed at process termination and result in any spawned processes also terminating    

## Acceptance criteria

* killing ibex-client.exe always terminates a spawned python.exe

## Notes

It may be possible to test this by temporarily modifying the code to call CloseHandle() on the handle returned from CreateJobObject() after setting the necessary option, this should immediately terminate the ibex-client process and any subprocesses

**Comment by NikolaRoev** [comment]
To reproduce issue:
- Run `sleep(n)` in Scripting.
- End the client process from the task manager.
- The pydev process is now orphaned.

As far as I can tell this only happens if the pydev process execution is suspended when the parent process is forcefully terminated. I was not able to reproduce it in any other way.

To reproduce the socket plotting error we can also do this:
- Plot something in a Scripting console.
- Plot something in another Scripting console.
- The socket we use in the matplotlib backend is in use by the previous instance so we get the exception.

Some solutions I've thought of:
- Rework this pydev function https://github.com/ISISComputingGroup/Pydev/blob/master/plugins/org.python.pydev.shared_core/src/org/python/pydev/shared_core/process/ProcessUtils.java#L104, and everything else it might affect, as outlined in this issue's description.
- Turn the matplotlib backend into a separate single instance process we can start with the other IBEX server components.
- Add a check to https://github.com/ISISComputingGroup/genie_python/blob/31632375f6bdad3f417e320a4553646866323bb8/Lib/site-packages/genie_python/matplotlib_backend/ibex_websocket_backend.py#L173 to see if the socket is currently in use, and if it is, stop the process using it.

**Comment by FreddieAkeroyd** [comment]
We may want to adopt several of the approaches, however I feel the job object approach is worth doing as it is a catch all cleanup that is wider than just matplotlib i.e. any hung spawned process (not just python/pydev) and file handles, global locks or other such things they may keep hold of if not terminated properly.

Option 3 (see if socket in use and terminate) if probabyl not needed with the job object fix, as per our other discussion on multiple clients trying to use the same plotting socket it may have the undesired effect of tryign to plot fm a second client kills off the pythin script running in the first console

Option 2 of a separate backend look a good additional future option to allow multiple clients to plot. However if we start this as a separate server component does that mean it will only work for ibex gui running on an instrument and not a remote connection from another pc? We may need to have it started once and only once per PC, it may be fine to let it terminate with the job object as the other client will restart it on the local PC, but we may need to be more careful (this is a bit how epics caRepeater works)

**Comment by FreddieAkeroyd** [comment]
I think the follow up issue should implement the job object and a new issue created to look at a  single instance process for matplotlib backend
